File: leetcode/product-of-array-except-self.py

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        """
        - Must be O(n) time
        - Can not use the division operation
        
        1. Create prefix and suffix arrays
            prefix[i] = product of nums[0...i-1]
            suffix[i] product if nums[i+1...n-1]
        2. Create the answer array, where
            answer[i] = prefix[i] * suffix[i]
            
        Time: O(n)
        Space O(n)
        """
        n = len(nums)
        prefix, suffix = [1]*n, [1]*n
        
        # prefix array
        for idx in range(1, n):
            prefix[idx] = prefix[idx-1]*nums[idx-1]
            
        # suffix array
        for idx in range(n-2, -1, -1):
            suffix[idx] = suffix[idx+1]*nums[idx+1]
            prefix[idx] *= suffix[idx]
        
        return prefix

File: leetcode/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> pre = vector<int>(nums.size());
        pre[0] = 0;
        pre[1] = nums[0];
        for (int i = 2; i < nums.size(); i++) {
            pre[i] = nums[i-1] * pre[i-1];
        }
        
        vector<int> suf = vector<int>(nums.size());
        suf[nums.size()-1] = 0;
        suf[nums.size()-2] = nums[nums.size()-1];
        for (int i = nums.size()-3; i >= 0; i--) {
            suf[i] = nums[i+1] * suf[i+1];
        }
        
        vector<int> res = vector<int>(nums.size());
        res[0] = suf[0];
        res[nums.size()-1] = pre[nums.size()-1];
        for (int i = 1; i < nums.size()-1; i++) {
            res[i] = pre[i] * suf[i];
        }
        
        return res;
    }
};
